File: backend/app/schemas/referral.py
from __future__ import annotations
import uuid
from datetime import datetime
from typing import Optional, List, Any
from pydantic import BaseModel, Field, field_validator
from app.models.referral import ReferralStatus, ArrivalCondition


class ReferralCreate(BaseModel):
    sex: str
    diagnosis: str
    reason_for_transfer: str
    # Which MoH transfer-form variant was used, and its form-specific field values
    # (a flat map keyed by field name — see the frontend transfer-form definitions).
    form_type: str = "EXTERNAL"
    form_data: Optional[dict[str, Any]] = None
    # The destination is required so a request is always routed to a specific
    # facility + unit (and only that side's staff can approve it).
    preferred_facility_id: uuid.UUID
    requested_unit_id: uuid.UUID
    # The resources being requested at the destination — one or more distinct
    # resources, each validated as currently available at the preferred facility
    # when creating the request.
    requested_resource_ids: List[uuid.UUID] = Field(min_length=1)
    # Optional voice-dictation artifacts, carried over from /referrals/transcribe
    # so the recording, transcript, and summary are stored with the referral.
    audio_url: Optional[str] = None
    transcript: Optional[str] = None
    ai_summary: Optional[str] = None
    # When the request follows a coordination call placed before the form was filled,
    # the call log's id — the new referral is linked back to it so both sides see it.
    call_log_id: Optional[uuid.UUID] = None


class ReferralDraftCreate(BaseModel):
    """A call-first "lightweight" referral: only the destination and requested
    resources are required. The detailed MoH transfer form (clinical fields and
    ``form_data``) is filled in later via ``ReferralUpdate``. Such a referral skips
    the in-app accept/reservation step — the phone call is the coordination — and
    goes straight to transport."""
    preferred_facility_id: uuid.UUID
    requested_unit_id: uuid.UUID
    requested_resource_ids: List[uuid.UUID] = Field(min_length=1)
    call_log_id: Optional[uuid.UUID] = None


class DictationFields(BaseModel):
    """Form fields extracted from a dictated transcript. Every field is optional —
    the clinician reviews and corrects before submitting."""
    sex: Optional[str] = None
    diagnosis: Optional[str] = None
    reason_for_transfer: Optional[str] = None


class DictationResult(BaseModel):
    """Response of /referrals/transcribe: the kept audio, the transcript, a short
    summary for the receiving clinic, and the extracted fields to prefill the form."""
    audio_url: Optional[str] = None
    transcript: str
    summary: str
    fields: DictationFields
    # Form-specific values extracted for the chosen MoH form variant (keyed by the
    # field names in the frontend transfer-form definitions). Empty when no form
    # spec was sent or nothing matched.
    form_data: dict[str, Any] = {}


class MonitoringVitalRow(BaseModel):
    """One vital-signs reading taken during transport (every ~30 minutes)."""
    time: Optional[str] = None
    bp: Optional[str] = None
    temp: Optional[str] = None
    spo2: Optional[str] = None
    rr: Optional[str] = None
    pulse: Optional[str] = None
    fhr: Optional[str] = None
    membranes_ruptured: Optional[str] = None


class MonitoringProblemRow(BaseModel):
    """A problem encountered during transport and how it was managed."""
    problem: Optional[str] = None
    management: Optional[str] = None


class TransportMonitoringResult(BaseModel):
    """The Patient Monitoring Transfer Form as recorded by the driver's voice:
    the kept recording, its transcript and summary, and the extracted log."""
    audio_url: Optional[str] = None
    transcript: str
    summary: str
    vital_signs: List[MonitoringVitalRow] = []
    problems: List[MonitoringProblemRow] = []
    recorded_at: Optional[datetime] = None


class ReferralUpdate(BaseModel):
    """Completing (or editing) the transfer form for a referral after creation —
    used to fill in the full MoH form for a call-first lightweight referral. Every
    field is optional; only what's provided is written."""
    sex: Optional[str] = None
    diagnosis: Optional[str] = None
    reason_for_transfer: Optional[str] = None
    form_type: Optional[str] = None
    form_data: Optional[dict[str, Any]] = None


class ReferralFeedbackRequest(BaseModel):
    """Receiving-side Referral Feedback and/or Counter-Referral. Either may be sent;
    each is a flat map keyed by the receiving-form field names."""
    feedback_data: Optional[dict[str, Any]] = None
    counter_referral_data: Optional[dict[str, Any]] = None


class AcceptReferralRequest(BaseModel):
    # Accepting reserves every resource the request asked for, so no resource is
    # chosen here — only the optional planned admission time applied to them all.
    planned_admission_time: Optional[datetime] = None


class RequestedResourceOut(BaseModel):
    """A resource named on a request, surfaced with its name so both facilities
    can see what was asked for (resolved server-side; the requesting side can't
    look up the destination's resources by id)."""
    id: uuid.UUID
    resource_name: str

    model_config = {"from_attributes": True}


class RejectReferralRequest(BaseModel):
    reason: str
    comment: Optional[str] = None


class ArrivalConditionRequest(BaseModel):
    arrival_condition: ArrivalCondition


class TransitStats(BaseModel):
    """Aggregate transit-duration stats (EN_ROUTE → ARRIVED), in minutes."""
    completed_journeys: int
    average_minutes: Optional[float] = None
    fastest_minutes: Optional[float] = None
    slowest_minutes: Optional[float] = None
    # Count of recorded patient arrival conditions, keyed by ArrivalCondition value.
    arrival_conditions: dict[str, int] = {}


class StatusHistoryOut(BaseModel):
    id: uuid.UUID
    status: ReferralStatus
    # Null when the change came from an ambulance driver (no staff user).
    changed_by: Optional[uuid.UUID] = None
    comment: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


class TransportEventRef(BaseModel):
    id: uuid.UUID
    ambulance_id: Optional[uuid.UUID] = None
    ambulance_identifier: str
    driver_name: Optional[str] = None
    driver_phone: Optional[str] = None
    dispatch_time: Optional[datetime] = None
    pickup_time: Optional[datetime] = None
    departure_time: Optional[datetime] = None
    arrival_time: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ReferralOut(BaseModel):
    id: uuid.UUID
    referral_number: str
    sex: str
    diagnosis: str
    reason_for_transfer: str
    form_type: str = "EXTERNAL"
    form_data: Optional[dict[str, Any]] = None
    form_completed: bool = True
    transport_monitorings: List[TransportMonitoringResult] = []
    feedback_data: Optional[dict[str, Any]] = None
    counter_referral_data: Optional[dict[str, Any]] = None
    audio_url: Optional[str] = None
    transcript: Optional[str] = None
    ai_summary: Optional[str] = None
    status: ReferralStatus
    rejection_reason: Optional[str]
    rejection_comment: Optional[str]
    arrival_condition: Optional[str] = None
    created_by: uuid.UUID
    referring_facility_id: Optional[uuid.UUID]
    preferred_facility_id: Optional[uuid.UUID]
    accepted_facility_id: Optional[uuid.UUID]
    origin_unit_id: Optional[uuid.UUID] = None
    requested_unit_id: Optional[uuid.UUID] = None
    requested_resources: List[RequestedResourceOut] = []
    # Which of the requested resources are actually reserved for this request — an
    # accept holds every one that was still available, so this can be a subset.
    reserved_resource_ids: List[uuid.UUID] = []
    created_at: datetime
    updated_at: datetime
    status_history: List[StatusHistoryOut] = []
    transport_events: List[TransportEventRef] = []

    @field_validator("transport_monitorings", mode="before")
    @classmethod
    def _monitorings_default(cls, v: Any) -> Any:
        # The column is null until the first recording; treat that as an empty list.
        return v or []

    model_config = {"from_attributes": True}


class ReferralSummary(BaseModel):
    id: uuid.UUID
    referral_number: str
    diagnosis: str
    status: ReferralStatus
    form_completed: bool = True
    created_at: datetime
    referring_facility_id: Optional[uuid.UUID]

    model_config = {"from_attributes": True}

File: frontend/src/types/referral.ts
export type ReferralStatus =
  | "DRAFT"
  | "REQUESTED"
  | "UNDER_REVIEW"
  | "ACCEPTED"
  | "TRANSPORT_ARRANGED"
  | "EN_ROUTE"
  | "ARRIVED"
  | "REJECTED"
  | "CANCELLED";

export type ArrivalCondition =
  | "STABLE"
  | "CRITICAL"
  | "DETERIORATED"
  | "ARRIVED_DECEASED";

export interface StatusHistory {
  id: string;
  status: ReferralStatus;
  changed_by: string;
  comment: string | null;
  created_at: string;
}

export interface ReferralTransport {
  id: string;
  ambulance_id: string | null;
  ambulance_identifier: string;
  driver_name: string | null;
  driver_phone: string | null;
  dispatch_time: string | null;
  pickup_time: string | null;
  departure_time: string | null;
  arrival_time: string | null;
}

/** A resource named on a request, with its name resolved server-side so both
 *  facilities can see what was asked for. */
export interface RequestedResource {
  id: string;
  resource_name: string;
}

/** The Patient Monitoring Transfer Form recorded by the ambulance driver by voice. */
export interface TransportMonitoring {
  audio_url: string | null;
  transcript: string;
  summary: string;
  vital_signs: Array<Record<string, string | null>>;
  problems: Array<Record<string, string | null>>;
  recorded_at: string | null;
}

export interface Referral {
  id: string;
  referral_number: string;
  sex: string;
  diagnosis: string;
  reason_for_transfer: string;
  form_type: string;
  form_data: Record<string, unknown> | null;
  /** False for a call-first lightweight referral whose full MoH form isn't filled in yet. */
  form_completed: boolean;
  transport_monitorings: TransportMonitoring[];
  feedback_data: Record<string, unknown> | null;
  counter_referral_data: Record<string, unknown> | null;
  audio_url: string | null;
  transcript: string | null;
  ai_summary: string | null;
  status: ReferralStatus;
  rejection_reason: string | null;
  rejection_comment: string | null;
  created_by: string;
  arrival_condition: string | null;
  referring_facility_id: string | null;
  preferred_facility_id: string | null;
  accepted_facility_id: string | null;
  origin_unit_id: string | null;
  requested_unit_id: string | null;
  requested_resources: RequestedResource[];
  reserved_resource_ids: string[];
  created_at: string;
  updated_at: string;
  status_history: StatusHistory[];
  transport_events: ReferralTransport[];
}

export interface CreateReferralPayload {
  sex: string;
  diagnosis: string;
  reason_for_transfer: string;
  form_type: string;
  form_data?: Record<string, unknown> | null;
  preferred_facility_id?: string;
  requested_unit_id?: string;
  requested_resource_ids: string[];
  audio_url?: string;
  transcript?: string;
  ai_summary?: string;
  /** Links a pre-form coordination call to the referral being created. */
  call_log_id?: string;
}

/** A call-first lightweight referral: only the destination + resources up front. */
export interface CreateDraftPayload {
  preferred_facility_id: string;
  requested_unit_id: string;
  requested_resource_ids: string[];
  call_log_id?: string;
}

/** Completing (or editing) the transfer form for a referral after creation. */
export interface CompleteReferralFormPayload {
  sex?: string;
  diagnosis?: string;
  reason_for_transfer?: string;
  form_type?: string;
  form_data?: Record<string, unknown> | null;
}

/** Form fields extracted from a dictated transcript (all optional — reviewed by the clinician). */
export interface DictationFields {
  sex?: string | null;
  diagnosis?: string | null;
  reason_for_transfer?: string | null;
}

/** Response of POST /referrals/transcribe. */
export interface DictationResult {
  audio_url: string | null;
  transcript: string;
  summary: string;
  fields: DictationFields;
  /** Form-specific values extracted for the chosen MoH form (keyed by field name). */
  form_data?: Record<string, unknown>;
}

export interface AcceptReferralPayload {
  planned_admission_time?: string;
}

export interface RejectReferralPayload {
  reason: string;
  comment?: string;
}
